### ISSUE TITLE
Replace voluptuous with a faster schema validator

### ISSUE THREAD

**Issue by ahal** [taskcluster/taskgraph#8]
Imported from: https://bugzilla.mozilla.org/show_bug.cgi?id=1652123

When profiling taskgraph in Gecko, we discovered that voluptuous is responsible for a significant performance hit to Taskgraph. So much so that we disbale schema validation entirely when `taskgraph.fast` is set. We also avoid using voluptuous to set default values and instead set defaults via a transform.

There are many more performant validation libraries out there. In another project, I once successfully replaced `voluptuous` with [validx](https://validx.readthedocs.io/en/latest/) (but we should investigate all options).

**Comment by bhearsum** [comment]
[pydantic](https://docs.pydantic.dev/latest/) seems to be the standard for this sort of thing these days. According to https://validx.readthedocs.io/en/latest/benchmarks.html, validx is ~5x/faster than pydantic 1.x. According to [this issues](https://github.com/pydantic/pydantic/discussions/6748#discussioncomment-12434390), pydantic 2.x is massively faster than 1.x when it comes to validation.

We should do a small scale test of validx vs. pydantic to help inform which we should choose.

**Comment by ahal** [comment]
Yeah, it looks like validx isn't well maintained anymore, so I'd lean towards pydantic between the two of them (even if 2.x is slightly slower).

I was also meaning to look into https://github.com/jcrist/msgspec, which I saw someone recommend as a faster, lighter weight alternative to pydantic.

**Comment by bhearsum** [comment]
https://www.reddit.com/r/Python/comments/w68nv3/a_quick_benchmark_comparing_pydantic_v2_with/ suggests that msgspec is significantly faster than even pydantic 2.x. https://gist.github.com/jcrist/d62f450594164d284fbea957fd48b743 has similar results (albeit that's from the msgspec author; I don't know if there's any bias there.)

**Comment by marco-c** [comment]
@ahal @bhearsum can we close this?

**Comment by jcristau** [comment]
Yes we can!
